Robot Framework pageobjects tutorial (#1089)
SFDO-Tooling/CumulusCI: docs/robotframework-tutorial.rst

========================
Robot Framework Tutorial
========================

> This tutorial will step you through writing your first test, then

Author: This question is answered in robotframework.rst, which comes just before this file in the generated documentation.

enhancing that test with a custom keyword implemented as a page
object. It is not a comprehensive tutorial on using robot

> framework. For robot framework documentations see the

Reviewer: Capitalize `Robot Framework` everywhere.

> Framework. For Robot Framework documentations see the

Reviewer 2: "documentation"

Author: fixed. Thanks for catching this.

`Robot Framework User Guide <http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html>`_

It is assumed you've worked through the CumulusCI :doc:`tutorial` at least up to the
point where you've called ``cci project init``.

Part 1: Folder Structure
========================

We recommend that all robot tests, keywords, data, and log and report files live under

> a folder named ``robot``, at the root of you repository. If you worked

Reviewer 2: "your"

Author: Fixed. Thanks for catching this.

through the :doc:`CumulusCI tutorial`, the following folders will
have been created under ``MyProject/robot/MyProject``:

- ``doc`` - a place to put documentation for your tests
- ``resources`` - a place to put robot libraries and keyword files that
  are unique to your project
- ``results`` - a place for robot to write its log and report files
- ``tests`` - a place for all of your tests. 

Part 2: Creating a custom object
================================

For this tutorial we're going to use a custom object named

> ``MyObject`` (eg: ``MyObject__c``). In addition, we need a custom tab that is associated

Reviewer: ```suggestion
``MyObject`` (e.g. ``MyObject__c``). In addition, we need a custom tab that is associated
```

with that object.

If you want to run the tests and keywords in this tutorial verbatim,
you will need to go to Setup and create the following:

1. A custom object with the name ``MyObject``.
2. A custom tab associated with this object.

Part 3: Creating and running your first robot test
==================================================

The first thing we want to do is create a test that verifies
we can get to the listing page of the custom object. This will

> let us know that everything is configure properly.

Reviewer 2: "configured"

Reviewer 2: @boakley I really like this. It adds a lot of depth and clarity to the existing documentation, as a Robot newbie. I highlighted just one spot that I thought was a little bit unclear above and flagged a few typos.

Author: Fixed. Thanks for catching this.

Open up your favorite editor and create a file named ``MyObject.robot``
in the folder ``robot/MyProject/tests``. Copy and paste the
following into this file, and then save it.

.. code-block:: console

    *** Settings ***
    Resource  cumulusci/robotframework/Salesforce.robot
    Library   cumulusci.robotframework.PageObjects

    Suite Setup     Open test browser
    Suite Teardown  Delete records and close browser

    *** Test Cases ***
    Test the MyObject listing page
        Go to page  Listing  MyObject__c
        Current page should be  Listing  MyObject__c

.. note::

>    The above code uses ``Go to page`` and ``Current page should be``

Author: The concept of keywords and page objects was introduced in the Robotframework section of the documentation, so I am not going to add anything here. I did, however, add a link at the very top of the tutorial to state that it's assumed the read has read that section before going through the tutorial.

    MyObject                                                                      
    ==============================================================================
    Test the MyObject listing page                                        | PASS |
    ------------------------------------------------------------------------------
    MyObject                                                              | PASS |
    1 critical test, 1 passed, 0 failed
    1 test total, 1 passed, 0 failed
    ==============================================================================
    Output:  /Users/boakley/dev/MyProject/robot/MyProject/results/output.xml
    Log:     /Users/boakley/dev/MyProject/robot/MyProject/results/log.html
    Report:  /Users/boakley/dev/MyProject/robot/MyProject/results/report.html

Part 4: Creating a page object
==============================

Most projects are going to need to write custom keywords that are
unique to that project. For example, NPSP has a keyword for filling in
a batch gift entry form, EDA has a keyword with some custom logic for
validating and affiliated contact, and so on.

The best way to create and organize these keywords is to place them in

> page object libraries. These libraries contain normal python classes

Reviewer: ```suggestion
page object libraries. These libraries contain normal Python classes
```

and methods which have been decorated with the ``pageobjects``
decorator provided by CumulusCI. By using page objects, you can write

> keywords that are unique to a given page, making them easier to find

Reviewer 3: I feel like there must be more to them than being easier to find and manage. Can you explain a little bit more what the page object infrastructure _does_ at runtime? Is it just an easier way of loading keywords? Or do the keywords have some magic associated with them by the decorator?

Author: This is a good comment.  You are right that maybe I haven't fully explained what page objects are here. They are introduced in the main robotframework section of the documentation, so I'm not going to add anything here. What I did do, however, was add a link at the very top of this file to encourage people to read that section before diving into the tutorial.

FWIW, the main benefit really is to make keywords easier to find and manage. That, and provides a bit more of a structured environment for creating keywords so that our keyword libraries aren't quite the wild wild west that they are right now.

Thanks for the comment!

> decorator provided by cumulusci. By using page objects, you can write

Reviewer: ```suggestion
decorator provided by CumulusCI. By using page objects, you can write
```

and easier to manage.

Defining the class
------------------

> cumulusci provides some base classes that are a good starting point

Reviewer: ```suggestion
CumulusCI provides some base classes that are a good starting point
```

for your page object. In this case we're writing a keyword that works
on the listing page, so we want our class to inherit from the

> ``ListingPage`` class.

Reviewer 2: (Oh, I see - it's in the other documentation file).

> ``ListingPage`` class.

Author: I've added a link from this paragraph to the one in the other document that mentions the different classes.

Our class also needs to use the ``pageobject`` decorator, so we must
import that along with the ``ListingPage`` class.

To get started, create a new file named ``MyObjectPages.py`` in the
folder ``robot/MyProject/resources``. At the top of the new keyword
file, add the following import statement: 

.. code-block:: python

    from cumulusci.robotframework.pageobjects import pageobject, ListingPage

Next we can create the class definition by adding the following two
lines:

.. code-block:: python

    @pageobject(page_type="Listing", object_name="MyObject__c")
    class MyObjectListingPage(ListingPage):

> The first line register this class as a page object for a listing page

Reviewer: ```suggestion
The first line registers this class as a page object for a listing page
```

characters ``...`` on a separate line.

Modify the import statements at the top of ``MyObject.robot`` to look
like the following:

.. code-block:: console

    *** Settings ***
    Resource  cumulusci/robotframework/Salesforce.robot
    Library   cumulusci.robotframework.PageObjects
    ...  robot/MyProject/resources/MyObjectPages.py

This will import the page object definitions into the test case, but
the keywords won't be available until the page object is loaded. Page
objects are loaded automatically when you call ``Go to page``, or you
can explicitly load them with ``Load page object``. In both cases, the
first argument is the page type (eg: `Listing`, `Home`, etc) and the
second argument is the object name (eg: ``MyObject__c``).

Our test is already using ``Go to page``, so our keyword should
already be available to us once we've gone to that page.

Part 5: Adding test data
========================

We want to be able to test that when we click on one of our custom
objects on the listing page that it will take us to the detail page
for that object. To do that, our test needs some test data. While that
can be very complicated in a real-world scenario, for simple tests we

> can use the salesforce API to create test data when the suite first

Reviewer 2: Cap "Salesforce"

Author: Fixed. Thanks for catching this.

right in the test to add some test data for us.

It is not necessary to do it in a setup. It could be a step in an
individual test case, for example. However, putting it in the ``Suite
Setup`` guarantees it will run before any tests in the same file are
run. 

Open up ``MyObject.robot`` and add the following just before ``***
Test Cases ***`:

.. code-block:: console

    *** Keywords ***
    Create test data
        [Documentation]
        ...  Creates a MyObject record named "Leeroy Jenkins"
        ...  if one doesn't exist

        # Check to see if the record is already in the database,
        # and return if it already exists
        ${status}  ${result}=  Run keyword and ignore error  Salesforce get  MyObject__c  Name=Leeroy Jenkins
        Return from keyword if  '${status}'=='PASS'

        # The record didn't exist, so create it
        Log  creating MyObject object with name 'Leeroy Jenkins'  DEBUG
        Salesforce Insert  MyObject__c  Name=Leeroy Jenkins

We also need to modify our ``Suite Setup`` to call this keyword in
addition to calling the ``Open Test Browser`` keyword. Since ``Suite
Setup`` only accepts a single keyword, we can use the built-in keyword

> `Run keywords` to run more than one keyword in the setup. 

Reviewer: check your consistency between double and single backticks.

Change the suite setup to look like the following, again using robot's
continuation characters to spread the code across multipe rows for
readability.

.. note::

    Note: it is critical that you use all caps for ``AND``, as
    that's the way robot knows where one keyword ends and the next
    begins.

.. code-block:: console

    Suite Setup     Run keywords
    ...  Create test data
    ...  AND  Open test browser

Notice that our ``Suite Teardown`` calls ``Delete records and
close browser``. The _records_ in that keyword refers to any data
records created by ``Salesforce Insert``. This makes it possible to
both create and later clean up temporary data used for a test.

It is important to note that the suite teardown isn't guaranteed to run
if you forcibly kill a running robot test. That is why we added the

> check in the ``Create test data`` keyword.

Reviewer 2: I don't quite follow this sentence. We added the check so that if the test is run twice against the same org it will not fail due to something like a duplicate rule?

Author: The scenario is this: the test adds "Leeroy Jenkins". If you forcibly stop the test before the teardown runs, that record will remain in the DB. The next time you run the test, without the check you'll end up with two records with that name. While essentially harmless in a scratch org, and harmless in this particular scenario, we don't want to be littering the database with unnecessary objects.

Author: I've reworded this paragraph slightly. Thanks for the comment!

Part 6: Using the new keyword
=============================

We are now ready to modify our test to use our new keyword, since we
now have some test data in our database, and the keyword definition in
our page object file.

Once again, edit ``MyObject.robot`` to add the following two
statements at the end of our test:

.. code-block:: console

    Click on the row with name  Leeroy Jenkins
    Current page should be  Detail  MyObject__c

The complete test should now look like this:

.. code-block:: console

    *** Settings ***
    Resource  cumulusci/robotframework/Salesforce.robot
    Library   cumulusci.robotframework.PageObjects
    ...  robot/MyProject/resources/MyObjectPages.py

    Suite Setup     Run keywords
    ...  Create test data
    ...  AND  Open test browser
    Suite Teardown  Delete records and close browser

    *** Keywords ***
    Create test data

>         [Documentation]  Creates a MyObject record named "Yoshi" if one doesn't exist

Reviewer: wrong name

        # Check to see if the record is already in the database,
        # and do nothing if it already exists
        ${status}  ${result}=  Run keyword and ignore error  Salesforce get  MyObject__c  Name=Leeroy Jenkins
        Return from keyword if  '${status}'=='PASS'

        # The record didn't exist, so create it
        Log  creating MyObject object with name 'Leeroy Jenkins'  DEBUG
        Salesforce Insert  MyObject__c  Name=Leeroy Jenkins

    *** Test Cases ***
    Test the MyObject listing page
        Go to page  Listing  MyObject__c
        Current page should be  Listing  MyObject__c

        Click on the row with name  Leeroy Jenkins
        Current page should be  Detail  MyObject__c

> With everthing in place, we should be able to run the test using the

Reviewer 2: "everything"

Author: fixed. Thanks for catching this.

        Salesforce Insert  MyObject__c  Name=Leeroy Jenkins

We also need to modify our ``Suite Setup`` to call this keyword in
addition to calling the ``Open Test Browser`` keyword. Since ``Suite
Setup`` only accepts a single keyword, we can use the built-in keyword
``Run keywords`` to run more than one keyword in the setup.

Change the suite setup to look like the following, again using robot's
continuation characters to spread the code across multipe rows for
readability.

.. note::

    It is critical that you use all caps for ``AND``, as
    that's the way robot knows where one keyword ends and the next
    begins.

.. code-block:: console

    Suite Setup     Run keywords
    ...  Create test data
    ...  AND  Open test browser

Notice that our ``Suite Teardown`` calls ``Delete records and
close browser``. The _records_ in that keyword refers to any data
records created by ``Salesforce Insert``. This makes it possible to
both create and later clean up temporary data used for a test.

It is important to note that the suite teardown isn't guaranteed to run
if you forcibly kill a running robot test. For that reason, we added a

> check in the ``Create test data`` to check for an existing record

Reviewer 4: ```suggestion
step in ``Create test data`` to check for an existing record
```
